Username typedtype filter

olivierh65 opened the pull request:
Filters formating drupal username from civicrm user:
FirstLastFilter (id : FirstLast) => EileenMcnaughton
FirstDotLastFilter (id : initialdotlast) => eileen.mcnaughton
InitialDotLastFilter (id : firstdotlast) => e.mcnaughton

Also add list of filters defined in civicrm_entity:
![image](https://user-images.githubusercontent.com/13754015/134020780-9d16249f-f23b-4ad4-9c7b-6f7ba27af514.png)

olivierh65 committed:
Create FirstLastFilter.php

olivierh65 committed:
Add 3 filters

Filters formating drupal username from civicrm user:
FirstLastFilter (id : FirstLast) => EileenMcnaughton
FirstDotLastFilter (id : initialdotlast) => eileen.mcnaughton
InitialDotLastFilter (id : firstdotlast) => e.mcnaughton

Usage example :
FORMAT
Format of the username. Use Twig style tokens for using the available data.
Value
{{civicrm_membership.contact_id.entity|initaldotlast}}

olivierh65 committed:
Add Civicrm filter list

olivierh65 committed:
Correct default values

Set default value to FALSE for notify and signin.
Change description for is_active.

olivierh65 committed:
Getdefault value if variable is empty

is_active, signin and notify default values are retrieved if variable is empty.

jackrabbithanna commented:
Very cool!!! Will be testing on test sites. Thanks for contributing!!
@puresyntax71 also let me know if you have any feedback

puresyntax71 commented:
Yeah, looks good to me. Just had some minor comment.

And perhaps shortening some lines like:

```php
  public function canFilter(DataDefinitionInterface $definition) {
    if ($definition->getConstraints()['EntityType'] == "civicrm_contact") {
        return true;
    }
    else {
        return false;
    }
  }
```

to

```php
  public function canFilter(DataDefinitionInterface $definition) {
    return $definition->getConstraints()['EntityType'] == "civicrm_contact";
  }
```

Yeah, very minor though. Don't think it's necessary to change it. :smiley:

jackrabbithanna commented on civicrm_entity.module:
```diff
@@ -561,5 +561,14 @@ function civicrm_entity_rebuild() {
+      $filters .= ($filters == '' ? '' : ', ') . $key['id'];
+    }
+  }
+  $rules_actions['civicrm_entity_user_create']['context']['format']->setDescription(t('Format of the username. Use <a href="@url">Twig style</a> tokens for using the available data.<br>Civicrm Entity filters available : @filters.',
```
instead of 'context', the array index will need to be 'context_definitions' .. That will need to be for this to work in Drupal 9

jackrabbithanna commented:
This PR will need rebasing before we can merge

olivierh65 committed:
adapt 'context' index name 

use 'context_definitions' index array if it contains index 'format' (Drupal 9), otherwise use 'context' (Drupal 8)

olivierh65 committed:
Merge branch '8.x-3.x' into Username-TypedTypeFilter

jackrabbithanna commented on civicrm_entity.module:
```diff
@@ -561,5 +561,22 @@ function civicrm_entity_rebuild() {
+      '@filters' => $filters]));
+  }
+  else {
+    $rules_actions['civicrm_entity_user_create']['context']['format']->setDescription(t('Format of the username. Use <a href="@url">Twig style</a> tokens for using the available data.<br>Civicrm Entity filter available : @filters.',
```
I think the rebase changed 'context_definitions' back to 'context' .. just that fix and will re-test to merge

olivierh65 committed:
Merge branch 'eileenmcnaughton:8.x-3.x' into Username-TypedTypeFilter
